server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -230,7 +233,9 @@ private void getFromTranslog(
         final var retryingListener = listener.delegateResponse((l, e) -> {
             final var cause = ExceptionsHelper.unwrapCause(e);
             logger.debug("get_from_translog failed", cause);
-            if (cause instanceof ShardNotFoundException || cause instanceof IndexNotFoundException) {
+            if (cause instanceof ShardNotFoundException
+                || cause instanceof IndexNotFoundException
+                || cause instanceof AlreadyClosedException) {
```

Comment:
I think the AlreadyClosedException should not use the cluster state observer, as there may be no new cluster state for the shard (when we unhollow, we don't create a new cluster state). It should call getFromTranslog() again, ideally in a retryable (with a few retries and some backoff time before it gives up).

Similar comment for MultiGet of course.

Reply:
Hmm, but if there's no cluster state change, the observer may timeout and fail the request. Maybe we should at least change onTimeout in case the cause was an ACE to try once more?

By the way, I think the ultimate solution would be ES-10826 and we'll wrap the whole operation around to guard against the engine closing.

Reply:
Good idea! I've verified that it works. Pushed in 4806d1133d08ddb708b03937de139a2f03aa33ed